File: kernel/epsilon/epsilon_core/thermodynamic_plasticity.py
# Epsilon-Hollow - Copyright (c) 2024 Teerth Sharma
# SPDX-License-Identifier: Epsilon-Hollow

"""
Epsilon-Hollow — Thermodynamic Erasure Bound for Liquid Plasticity (TEB)
==========================================================================
Author: Teerth Sharma
Status: Original Theorem — Lean 4 Verification Pending

═══════════════════════════════════════════════════════════════
THEOREM 8: Thermodynamic Erasure Bound (TEB)
═══════════════════════════════════════════════════════════════

Connects Landauer's principle from thermodynamics to the Liquid Tensor's
hot partition. Gives a physically fundamental floor on the energy cost
of continuous learning.

Background (Landauer's Principle, 1961):
    Erasing one bit of information dissipates at least k_B T ln(2) joules,
    where k_B ≈ 1.381 × 10⁻²³ J/K is Boltzmann's constant and T is
    temperature in Kelvin.

    This is not an engineering limit — it's a consequence of the second
    law of thermodynamics. No computer can do better.

Definition (Weight Erasure):
    Each micro-gradient update to the hot partition overwrites |W_hot|
    weights. In information-theoretic terms, each float64 weight stores
    ~52 bits of mantissa information. Overwriting = erasing + rewriting.

    Total bits erased per update:
        B_erase = |W_hot| × b_precision

    where b_precision = 52 for float64, 23 for float32, 10 for float16.

Theorem (Minimum Energy per Learning Step):
    The Liquid Tensor with hot partition W_hot dissipates at least:

        E_min = |W_hot| × b_precision × k_B × T × ln(2)

    joules per inject_update() call.

    Proof:
        1. Each update overwrites |W_hot| weights.
        2. Each weight carries b_precision bits of information.
        3. By Landauer: erasing each bit costs ≥ k_B T ln(2).
        4. Total: E_min = |W_hot| × b_precision × k_B T ln(2).

    ∎

Corollary (Optimal Hot Ratio):
    Given a power budget P_budget (watts) and update frequency f (Hz),
    the maximum sustainable hot ratio is:

        r_max = P_budget / (f × |W| × b_precision × k_B × T × ln(2))

    For the Epsilon-Hollow defaults (r = 0.005):
        With |W| = 70 × 10⁹ (70B model), b = 16 (fp16), T = 300K, f = 20 Hz:
        P_min = 0.005 × 70×10⁹ × 16 × 1.381×10⁻²³ × 300 × ln(2) × 20
              ≈ 6.4 × 10⁻⁸ watts

    This is 12 orders of magnitude below actual GPU power (~300W),
    confirming the hot ratio is WAY above the thermodynamic floor.
    The real bottleneck is compute, not thermodynamics.

Theorem (Information-Theoretic Learning Rate):
    The entropy-adaptive learning rate η = α(1 − H(P)) can be
    rewritten in thermodynamic terms:

        η_thermo = α × (S_max − S) / S_max

    where S = −k_B Σ pᵢ ln(pᵢ) is the Gibbs entropy of the output
    distribution. When S → 0 (minimum entropy), η → α (maximum
    learning). When S → S_max (maximum entropy), η → 0 (no learning).

    This is equivalent to the "free energy" formulation:
        η ∝ −ΔF / k_B T

    where ΔF = ΔU − TΔS is the Helmholtz free energy change.
    Learning occurs when ΔF < 0 (thermodynamically favorable).

    Proof:
        1. Shannon entropy H(P) = −Σ pᵢ log₂(pᵢ)
        2. Gibbs entropy S = k_B ln(2) × H(P) (conversion factor)
        3. S_max = k_B ln(|P|) = k_B ln(2) × log₂(|P|) = k_B ln(2) × H_max
        4. η = α(1 − H/H_max) = α(1 − S/(k_B ln(2) H_max))
             = α(S_max − S)/S_max
        5. The free energy interpretation:
           ΔF = (internal energy change) − T × (entropy change)
           When S < S_max: ΔF < 0 → favorable → learn
           When S = S_max: ΔF = 0 → equilibrium → don't learn

    ∎

═══════════════════════════════════════════════════════════════
"""
from __future__ import annotations

import math
from typing import Any, Dict, Optional

import numpy as np

# Physical constants
K_BOLTZMANN = 1.380649e-23  # J/K (exact, SI 2019)
LN2 = math.log(2)
LANDAUER_LIMIT = K_BOLTZMANN * LN2  # J/bit at T=1K


def landauer_energy_per_bit(temperature_K: float = 300.0) -> float:
    """
    Minimum energy to erase one bit of information.

    E_bit = k_B × T × ln(2)

    At room temperature (300K): ≈ 2.87 × 10⁻²¹ J/bit
    """
    return K_BOLTZMANN * temperature_K * LN2


def min_energy_per_update(n_hot_params: int, precision_bits: int = 52,
                          temperature_K: float = 300.0) -> float:
    """
    Minimum energy per inject_update() call.

    E_min = |W_hot| × b_precision × k_B × T × ln(2)

    Parameters
    ----------
    n_hot_params : int
        Number of weights in the hot partition.
    precision_bits : int
        Bits per weight (52 for float64, 23 for float32, 10 for float16).
    temperature_K : float
        Operating temperature in Kelvin.
    """
    e_bit = landauer_energy_per_bit(temperature_K)
    return n_hot_params * precision_bits * e_bit


def max_sustainable_hot_ratio(total_params: int, power_budget_watts: float,
                              update_freq_hz: float,
                              precision_bits: int = 16,
                              temperature_K: float = 300.0) -> float:
    """
    Maximum hot ratio sustainable within a power budget.

    r_max = P_budget / (f × |W| × b × k_B × T × ln(2))
    """
    e_bit = landauer_energy_per_bit(temperature_K)
    denom = update_freq_hz * total_params * precision_bits * e_bit
    if denom < 1e-30:
        return float("inf")
    return power_budget_watts / denom


def gibbs_entropy(distribution: np.ndarray) -> float:
    """
    Gibbs entropy: S = −k_B Σ pᵢ ln(pᵢ)

    Related to Shannon entropy by: S = k_B × ln(2) × H(P)
    """
    p = np.abs(distribution) + 1e-30
    p = p / np.sum(p)
    return -K_BOLTZMANN * float(np.sum(p * np.log(p)))


def thermodynamic_lr(alpha: float, gibbs_S: float,
                     S_max: float) -> float:
    """
    η_thermo = α × (S_max − S) / S_max

    Thermodynamic learning rate from Gibbs entropy.
    """
    if S_max < 1e-30:
        return alpha
    return alpha * max(0.0, (S_max - gibbs_S) / S_max)


def helmholtz_free_energy_change(delta_U: float, temperature_K: float,
                                 delta_S: float) -> float:
    """
    ΔF = ΔU − TΔS

    Learning is favorable when ΔF < 0.
    """
    return delta_U - temperature_K * delta_S


class ThermodynamicAnalyzer:
    """
    Analyzes the thermodynamic limits of the Liquid Tensor.

    Computes:
        1. Landauer power floor for continuous learning
        2. Comparison to actual GPU power consumption
        3. Information-theoretic bounds on learning rate
        4. Free energy favorability of updates
    """

    def __init__(self, total_params: int = 70_000_000_000,
                 hot_ratio: float = 0.005,
                 precision_bits: int = 16,
                 temperature_K: float = 300.0,
                 gpu_power_watts: float = 700.0):  # H100 SXM TDP
        self.total_params = total_params
        self.hot_ratio = hot_ratio
        self.hot_params = int(total_params * hot_ratio)
        self.precision_bits = precision_bits
        self.temperature_K = temperature_K
        self.gpu_power = gpu_power_watts

    def energy_analysis(self, update_freq_hz: float = 20.0) -> Dict[str, Any]:
        """Full energy analysis."""
        e_per_update = min_energy_per_update(
            self.hot_params, self.precision_bits, self.temperature_K
        )
        power_floor = e_per_update * update_freq_hz

        r_max = max_sustainable_hot_ratio(
            self.total_params, self.gpu_power, update_freq_hz,
            self.precision_bits, self.temperature_K
        )

        # How many orders of magnitude above Landauer?
        if power_floor > 0:
            orders_above = math.log10(self.gpu_power / power_floor)
        else:
            orders_above = float("inf")

        return {
            "hot_params": self.hot_params,
            "precision_bits": self.precision_bits,
            "temperature_K": self.temperature_K,
            "landauer_per_bit_J": landauer_energy_per_bit(self.temperature_K),
            "energy_per_update_J": e_per_update,
            "power_floor_watts": power_floor,
            "gpu_power_watts": self.gpu_power,
            "orders_above_landauer": orders_above,
            "max_hot_ratio_thermodynamic": r_max,
            "current_hot_ratio": self.hot_ratio,
            "thermodynamic_headroom": r_max / self.hot_ratio if self.hot_ratio > 0 else float("inf"),
        }

    def lr_analysis(self, output_distribution: np.ndarray) -> Dict[str, Any]:
        """
        Compare Shannon and Gibbs formulations of entropy-adaptive lr.
        """
        # Shannon
        p = np.abs(output_distribution) + 1e-30
        p = p / np.sum(p)
        H_shannon = -float(np.sum(p * np.log2(p)))
        H_max_shannon = math.log2(len(p))

        # Gibbs
        S_gibbs = gibbs_entropy(output_distribution)
        S_max = K_BOLTZMANN * math.log(len(p))

        # Learning rates
        alpha = 0.01
        eta_shannon = alpha * (1.0 - H_shannon / max(H_max_shannon, 1e-10))
        eta_gibbs = thermodynamic_lr(alpha, S_gibbs, S_max)

        # Free energy
        delta_U = 0.001  # Assume small internal energy change
        delta_S = S_max - S_gibbs
        delta_F = helmholtz_free_energy_change(
            delta_U, self.temperature_K, delta_S
        )

        return {
            "H_shannon": H_shannon,
            "H_max": H_max_shannon,
            "S_gibbs": S_gibbs,
            "S_max": S_max,
            "eta_shannon": eta_shannon,
            "eta_gibbs": eta_gibbs,
            "lr_agreement": abs(eta_shannon - eta_gibbs) < 1e-10,
            "delta_F": delta_F,
            "learning_favorable": delta_F < 0,
        }

    def h100_cluster_analysis(self, n_gpus: int = 8) -> Dict[str, Any]:
        """
        Multi-H100 DGX cluster analysis.

        8× H100 SXM: 8 × 700W = 5.6 kW GPU power.
        """
        total_power = self.gpu_power * n_gpus
        energy = self.energy_analysis()

        # At cluster scale
        cluster_r_max = max_sustainable_hot_ratio(
            self.total_params, total_power, 20.0,
            self.precision_bits, self.temperature_K
        )

        return {
            "n_gpus": n_gpus,
            "total_gpu_power_kw": total_power / 1000,
            "landauer_floor_watts": energy["power_floor_watts"],
            "orders_above_landauer": energy["orders_above_landauer"],
            "max_hot_ratio_cluster": cluster_r_max,
            "practical_limit": "Compute-bound, not thermodynamic-bound",
            "conclusion": (
                f"The hot ratio {self.hot_ratio} is {energy['thermodynamic_headroom']:.1e}× "
                f"above the Landauer floor. Epsilon-Hollow's continuous learning "
                f"is thermodynamically viable at H100 scale with "
                f"{energy['orders_above_landauer']:.0f} orders of magnitude headroom."
            ),
        }

File: kernel/epsilon/epsilon/crates/aether-core/src/thermodynamic_plasticity.rs
// Epsilon-Hollow - Copyright (c) 2024 Teerth Sharma
// SPDX-License-Identifier: Epsilon-Hollow

//! Thermodynamic Erasure Bound (TEB) formulas in `no_std` Rust.
//!
//! This module replaces the legacy `epsilon_core/thermodynamic_plasticity.py`
//! theorem slice. It keeps the Landauer energy floor, hot-partition power
//! bound, Gibbs entropy learning-rate equivalence, and Helmholtz free-energy
//! checks available to the Rust/Aether theorem runtime without NumPy or Python.

use libm::{fabs, log, log10};

/// Boltzmann constant in joules per kelvin.
pub const K_BOLTZMANN: f64 = 1.380_649e-23;
/// Natural logarithm of 2.
pub const LN2: f64 = core::f64::consts::LN_2;
/// Landauer limit in joules per bit at 1 kelvin.
pub const LANDAUER_LIMIT: f64 = K_BOLTZMANN * LN2;
/// Default total parameter count for the H100 theorem slice.
pub const H100_TOTAL_PARAMS: u64 = 70_000_000_000;
/// Default hot partition ratio for continuous learning.
pub const H100_HOT_RATIO: f64 = 0.005;
/// Default precision bits for fp16 hot weights.
pub const H100_PRECISION_BITS: u32 = 16;
/// Default operating temperature in kelvin.
pub const ROOM_TEMPERATURE_K: f64 = 300.0;
/// H100 SXM thermal design power used by the legacy theorem slice.
pub const H100_GPU_POWER_WATTS: f64 = 700.0;
const ENTROPY_EPS: f64 = 1e-30;

/// Minimum energy to erase one bit at `temperature_k`.
pub fn landauer_energy_per_bit(temperature_k: f64) -> f64 {
    if !temperature_k.is_finite() || temperature_k <= 0.0 {
        return 0.0;
    }
    K_BOLTZMANN * temperature_k * LN2
}

/// Minimum energy per hot-partition update.
pub fn min_energy_per_update(n_hot_params: u64, precision_bits: u32, temperature_k: f64) -> f64 {
    n_hot_params as f64 * f64::from(precision_bits) * landauer_energy_per_bit(temperature_k)
}

/// Maximum hot ratio sustainable inside a power budget.
pub fn max_sustainable_hot_ratio(
    total_params: u64,
    power_budget_watts: f64,
    update_freq_hz: f64,
    precision_bits: u32,
    temperature_k: f64,
) -> f64 {
    let denom = update_freq_hz
        * total_params as f64
        * f64::from(precision_bits)
        * landauer_energy_per_bit(temperature_k);
    if !denom.is_finite() || denom < ENTROPY_EPS {
        return f64::INFINITY;
    }
    power_budget_watts / denom
}

/// Gibbs entropy `S = -k_B * sum(p_i ln p_i)` for a signed distribution slice.
pub fn gibbs_entropy(distribution: &[f64]) -> f64 {
    let total = abs_sum_with_floor(distribution);
    if total <= 0.0 {
        return 0.0;
    }

    let mut entropy = 0.0;
    for value in distribution {
        let p = (fabs(*value) + ENTROPY_EPS) / total;
        entropy -= K_BOLTZMANN * p * log(p);
    }
    entropy
}

/// Entropy-adaptive learning rate in thermodynamic form.
pub fn thermodynamic_lr(alpha: f64, gibbs_s: f64, s_max: f64) -> f64 {
    if !s_max.is_finite() || s_max < ENTROPY_EPS {
        return alpha;
    }
    let ratio = (s_max - gibbs_s) / s_max;
    alpha * ratio.max(0.0)
}

/// Helmholtz free-energy change `delta_F = delta_U - T * delta_S`.
pub fn helmholtz_free_energy_change(delta_u: f64, temperature_k: f64, delta_s: f64) -> f64 {
    delta_u - temperature_k * delta_s
}

/// Analyzer configuration for TEB energy and learning-rate reports.
#[derive(Debug, Clone, Copy, PartialEq)]
pub struct ThermodynamicAnalyzer {
    /// Total model parameter count.
    pub total_params: u64,
    /// Ratio of parameters in the hot mutable partition.
    pub hot_ratio: f64,
    /// Number of hot mutable parameters.
    pub hot_params: u64,
    /// Precision bits per hot parameter.
    pub precision_bits: u32,
    /// Operating temperature in kelvin.
    pub temperature_k: f64,
    /// Available GPU power in watts.
    pub gpu_power_watts: f64,
}

impl ThermodynamicAnalyzer {
    /// Create an analyzer from explicit model and hardware parameters.
    pub fn new(
        total_params: u64,
        hot_ratio: f64,
        precision_bits: u32,
        temperature_k: f64,
        gpu_power_watts: f64,
    ) -> Self {
        let bounded_ratio = if hot_ratio.is_finite() && hot_ratio > 0.0 {
            hot_ratio
        } else {
            0.0
        };
        Self {
            total_params,
            hot_ratio: bounded_ratio,
            hot_params: (total_params as f64 * bounded_ratio) as u64,
            precision_bits,
            temperature_k,
            gpu_power_watts,
        }
    }

    /// Create the default H100/70B analyzer used by the legacy theorem slice.
    pub fn h100_default() -> Self {
        Self::new(
            H100_TOTAL_PARAMS,
            H100_HOT_RATIO,
            H100_PRECISION_BITS,
            ROOM_TEMPERATURE_K,
            H100_GPU_POWER_WATTS,
        )
    }

    /// Compute the Landauer floor and headroom for a given update frequency.
    pub fn energy_analysis(&self, update_freq_hz: f64) -> EnergyAnalysis {
        let energy_per_update =
            min_energy_per_update(self.hot_params, self.precision_bits, self.temperature_k);
        let power_floor_watts = energy_per_update * update_freq_hz;
        let max_hot_ratio_thermodynamic = max_sustainable_hot_ratio(
            self.total_params,
            self.gpu_power_watts,
            update_freq_hz,
            self.precision_bits,
            self.temperature_k,
        );
        let orders_above_landauer = if power_floor_watts > 0.0 {
            log10(self.gpu_power_watts / power_floor_watts)
        } else {
            f64::INFINITY
        };
        let thermodynamic_headroom = if self.hot_ratio > 0.0 {
            max_hot_ratio_thermodynamic / self.hot_ratio
        } else {
            f64::INFINITY
        };

        EnergyAnalysis {
            hot_params: self.hot_params,
            precision_bits: self.precision_bits,
            temperature_k: self.temperature_k,
            landauer_per_bit_j: landauer_energy_per_bit(self.temperature_k),
            energy_per_update_j: energy_per_update,
            power_floor_watts,
            gpu_power_watts: self.gpu_power_watts,
            orders_above_landauer,
            max_hot_ratio_thermodynamic,
            current_hot_ratio: self.hot_ratio,
            thermodynamic_headroom,
        }
    }

    /// Compare Shannon and Gibbs formulations of entropy-adaptive learning.
    pub fn lr_analysis(&self, output_distribution: &[f64], alpha: f64) -> LearningRateAnalysis {
        let h_shannon = shannon_entropy(output_distribution);
        let h_max = if output_distribution.len() > 1 {
            log(output_distribution.len() as f64) / LN2
        } else {
            0.0
        };
        let s_gibbs = gibbs_entropy(output_distribution);
        let s_max = if output_distribution.len() > 1 {
            K_BOLTZMANN * log(output_distribution.len() as f64)
        } else {
            0.0
        };
        let eta_shannon = if h_max > ENTROPY_EPS {
            alpha * (1.0 - h_shannon / h_max).max(0.0)
        } else {
            alpha
        };
        let eta_gibbs = thermodynamic_lr(alpha, s_gibbs, s_max);
        let delta_s = s_max - s_gibbs;
        let delta_f = helmholtz_free_energy_change(0.001, self.temperature_k, delta_s);

        LearningRateAnalysis {
            h_shannon,
            h_max,
            s_gibbs,
            s_max,
            eta_shannon,
            eta_gibbs,
            lr_agreement: fabs(eta_shannon - eta_gibbs) < 1e-10,
            delta_f,
            learning_favorable: delta_f < 0.0,
        }
    }

    /// Analyze a multi-H100 cluster with `n_gpus` devices.
    pub fn h100_cluster_analysis(&self, n_gpus: u32) -> ClusterAnalysis {
        let energy = self.energy_analysis(20.0);
        let total_power = self.gpu_power_watts * f64::from(n_gpus);
        let max_hot_ratio_cluster = max_sustainable_hot_ratio(
            self.total_params,
            total_power,
            20.0,
            self.precision_bits,
            self.temperature_k,
        );

        ClusterAnalysis {
            n_gpus,
            total_gpu_power_kw: total_power / 1000.0,
            landauer_floor_watts: energy.power_floor_watts,
            orders_above_landauer: energy.orders_above_landauer,
            max_hot_ratio_cluster,
            practical_limit: "Compute-bound, not thermodynamic-bound",
        }
    }
}

/// Landauer floor report for a hot-partition update loop.
#[derive(Debug, Clone, Copy, PartialEq)]
pub struct EnergyAnalysis {
    /// Number of hot mutable parameters.
    pub hot_params: u64,
    /// Precision bits per hot parameter.
    pub precision_bits: u32,
    /// Operating temperature in kelvin.
    pub temperature_k: f64,
    /// Landauer floor per bit in joules.
    pub landauer_per_bit_j: f64,
    /// Minimum energy per update in joules.
    pub energy_per_update_j: f64,
    /// Minimum continuous update power in watts.
    pub power_floor_watts: f64,
    /// Available GPU power in watts.
    pub gpu_power_watts: f64,
    /// Base-10 orders of magnitude between GPU power and Landauer floor.
    pub orders_above_landauer: f64,
    /// Maximum hot ratio allowed by the power budget.
    pub max_hot_ratio_thermodynamic: f64,
    /// Current configured hot ratio.
    pub current_hot_ratio: f64,
    /// Ratio between thermodynamic maximum and configured hot ratio.
    pub thermodynamic_headroom: f64,
}

/// Learning-rate equivalence report for Shannon and Gibbs entropy.
#[derive(Debug, Clone, Copy, PartialEq)]
pub struct LearningRateAnalysis {
    /// Shannon entropy in bits.
    pub h_shannon: f64,
    /// Maximum Shannon entropy in bits.
    pub h_max: f64,
    /// Gibbs entropy in joules per kelvin.
    pub s_gibbs: f64,
    /// Maximum Gibbs entropy in joules per kelvin.
    pub s_max: f64,
    /// Entropy-adaptive learning rate via Shannon entropy.
    pub eta_shannon: f64,
    /// Entropy-adaptive learning rate via Gibbs entropy.
    pub eta_gibbs: f64,
    /// True when both learning-rate formulations agree numerically.
    pub lr_agreement: bool,
    /// Helmholtz free-energy change.
    pub delta_f: f64,
    /// True when the configured free-energy change is favorable.
    pub learning_favorable: bool,
}

/// H100 cluster-scale thermodynamic report.
#[derive(Debug, Clone, Copy, PartialEq)]
pub struct ClusterAnalysis {
    /// Number of H100 GPUs in the cluster.
    pub n_gpus: u32,
    /// Total GPU power in kilowatts.
    pub total_gpu_power_kw: f64,
    /// Landauer continuous-update floor in watts.
    pub landauer_floor_watts: f64,
    /// Orders of magnitude between GPU power and Landauer floor.
    pub orders_above_landauer: f64,
    /// Maximum hot ratio allowed by total cluster power.
    pub max_hot_ratio_cluster: f64,
    /// Practical bottleneck classification.
    pub practical_limit: &'static str,
}

fn shannon_entropy(distribution: &[f64]) -> f64 {
    let total = abs_sum_with_floor(distribution);
    if total <= 0.0 {
        return 0.0;
    }

    let mut entropy = 0.0;
    for value in distribution {
        let p = (fabs(*value) + ENTROPY_EPS) / total;
        entropy -= p * (log(p) / LN2);
    }
    entropy
}

fn abs_sum_with_floor(distribution: &[f64]) -> f64 {
    if distribution.is_empty() {
        return 0.0;
    }

    let mut total = 0.0;
    for value in distribution {
        total += fabs(*value) + ENTROPY_EPS;
    }
    total
}
